Expand Contributor's Guide and README (#52)
NCAR/geocat-applications: CONTRIBUTING.md

```
git config --global user.email "<username>@users.noreply.github.com"
```
Verify the email has been set correctly
```
git config --global user.email
```
If the command prints out the email provided then the email has been set correctly.

### Setup Environment

First, [install Miniconda](https://docs.anaconda.com/free/miniconda/miniconda-install/).
Conda is the command line interface for Miniconda and is a useful tool to manage environments
and dependencies. A conda environment is created from the `environment.yml` that contains
a list of required dependencies.

```
# Create a new conda environment with required dependencies
conda env create -f environment.yml

# Activate your new environment
conda activate geocat-applications
```
If the environment has been activated correctly, then the terminal will be preceded by:

```
(geocat-applications) user@user-os:~$
```

To view all conda environments created locally:
```
conda env list
```

### Create Branch for Changes
We recommend creating a new branch on your fork reposisitory for the new feature or bug fix.

To create a new branch: (replace `<new-branch-name>` with your new branch name)
```
git checkout -b <new-branch-name>

> ```

Reviewer: Gosh, github markdown doesn't like that. To be clear, my suggestion is 
````
To create a new branch (replace `<new-branch-name>` with your new branch name):
```
git checkout -b <new-branch-name>
```
````

# Contributor Guide
Welcome to `geocat-applications`! Thank you for your interest in contributing to this project! This guide
describes how to contribute to `geocat-applications` and help us expand this reference for others.

If you have any questions, you can also reach us by email at [geocat@ucar.edu](geocat@ucar.edu)

## How to Contribute

There are many ways to contribute:

- [Proofreading and reporting bugs](https://github.com/NCAR/geocat-applications/issues/new/choose) 🐞
- [Requesting new features](https://github.com/NCAR/geocat-applications/issues/new/choose) 💡

> - Contribute a new [NCL to Python](#ncl-entry) page

Reviewer 2: ```suggestion
- Contributing a new [NCL to Python](#ncl-entry) page
```

> - Contribute a new [Python Entry](#python-entry) page

Reviewer 2: ```suggestion
- Contributing a new [Python Entry](#python-entry) page
```

For consistency

## Development Workflow Overview
This is the general development workflow to create a new page, edit, or
expand on an existing page:

**Setting up Repo and Local Development Environment**

1. [Setup GitHub and Repo](#Setup-Github-and-Repo)
2. [Setup the Environment](#Setup-the-Environment)
3. [Install Pre-Commit Hooks](#Install-Pre-Commit-Hooks)

**Making Your Changes**

1. [Understanding the Repository](#Understanding-the-Repository)

> 2. [Types of `geocat-applications` pages](#Types-of-geocat-applications-pages)

Reviewer 2: ```suggestion
2. [Types of `geocat-applications` Pages](#Types-of-geocat-applications-pages)
```

- [Proofreading and reporting bugs](https://github.com/NCAR/geocat-applications/issues/new/choose) 🐞
- [Requesting new features](https://github.com/NCAR/geocat-applications/issues/new/choose) 💡
- Contribute a new [NCL to Python](#ncl-entry) page
- Contribute a new [Python Entry](#python-entry) page

## Development Workflow Overview
This is the general development workflow to create a new page, edit, or
expand on an existing page:

**Setting up Repo and Local Development Environment**

1. [Setup GitHub Account and Fork Repo](#creating-a-free-github-account)
2. [Setup Environment](#Setup-Environment)
3. [Create Branch for Changes](#Create-Branch-for-Changes)
4. [Install Pre-Commit Hooks](#Install-Pre-Commit-Hooks)

**Making Your Changes**

1. [Understanding the Repository](#Understanding-the-Repository)
2. [Types of `geocat-applications` pages](#Types-of-geocat-applications-pages)
3. [Generate the Documentation Locally](#Generate-the-Documentation-Locally)

**Contribute Code and Review**

1. [Check Files Changed](#Check-Files-Changed)
2. [Open a New Pull Request](#Open-a-New-Pull-Request)
3. [Address Feedback](#Address-Feedback)
4. [Delete Branch](#Delete-Branch)

> ## Setting up Repo and Local Development Environment

Reviewer 2: A good bit of this section is intro content covered by Project Pythia Foundations.  I wonder if we should point to that for folks who are new to Git / GitHub rather than detailing this here.

Author: It would be useful to point people to Pythia as further resources, but I think it would get confusing to be jumping back and forth between documentation when you only need a few of the steps to get this done for applications

Reviewer 2: My main concern is that if you haven't worked with git at all or are unfamiliar with GitHub this section is probably not enough and some terminology used here may be confusing (e.g. locally, fork, repo) and if you have even a bit of familiarity with these tools this section is probably too much.  I think recommending something like the Project Pythia Foundations chapter and then keeping things a little tighter here gives people more of what they need and gets the git installation and configuration out of the way.

If you feel strongly about keeping it as is, I think we should at least adjust the language a bit and add some links to Project Pythia and/or Software Carpentry.

Reviewer: I can see both sides here, but ultimately as long as we convey that we want people to fork our project to contribute, I'll be okay.

1. Make sure to clear and run all outputs before asking for a review

[Example of a NCL Entry](https://ncar.github.io/geocat-applications/ncl/ncl_entries/days_in_month.html)

#### Receipts

Receipt files are small files with little to no narrative content that are for
the purpose of adding an entry to the NCL Index without the need for a full
**NCL Entry**. These files are accessible from the geocat-applications webpage,
but are not listed on TOCs or intended to be read as standalone or
comprehensive guides. They are intended to provide more extensive testing than is useful
in an NCL entry. Receipts are intended to be the minimal amount of documentation necessary to
add an entry to the NCL Index in cases where a full **NCL Entry** is not
necessary or where providing an initial entry to the NCL Index is more important
than waiting for a full **NCL Entry** to be completed.

1. Create a new file in `ncl/ncl_receipts/` based off of the
   `templates/receipt_template.ipynb` template.

> 1. Create raw `.ncl` script within `ncl/ncl_raw/` to store `NCL Code` section

Reviewer 2: ```suggestion
1. Create a raw `.ncl` script within `ncl/ncl_raw/` to store `NCL Code` section
```

1. Remove placeholder and descriptive text below `NCL Code`,
`Python Functionality`, `Comparison`

1. Make sure to clear and run all outputs before asking for a review

1. Add a new line to the `ncl/ncl_index/ncl-index-table.csv` file

[Example of a NCL Receipt](https://ncar.github.io/geocat-applications/ncl/receipts/days_in_month.html)

### Generate the Documentation Locally
From the `geocat-applications/` directory, run:
```
make clean html
```
This will build the documentation locally so you can view your changes.

New documentation can be viewed by opening `index.html` generated under
 `_build/html/index.html` on a local browser:

```

> # on Mac

Reviewer 2: ```suggestion
# on macOS
```

open _build/html/index.html

# Otherwise, open with a specific browser, like Firefox
firefox _build/html/index.html
```
Running these commands from the terminal will open a new browser tab with the
changes you have made to the repository. The webpage runs on your local host,
but all the pages are linked together and act as they will on the
`geocat-applications` website. You can browse your changes, navigate with links,
and view images as you would on the official website.

## Contribute the Code
Once you have made your changes on notebooks and they are ready for review

> by the GeoCAT team, you can open a new Pull Request. This section describes how

Reviewer 2: ```suggestion
by the GeoCAT team, you can open a new pull request. This section describes how
```

to open a pull request and what to expect after you open it.

### Check Files Changed

For a Python Entry, the files changed should include:
- New Python page under relevant `applications` folder
- Link to notebook in `.rst` file in the same directory as `applications/` (for
example: if new notebook is part of `data_analysis`, add link to
`applications/data_analysis/data_analysis.rst`)
- Link new notebook in `applications/applications.rst` file

For a NCL to Python entry, the files changes should include:
- New NCL to Python notebook under `ncl/ncl_entries`
- Link to new `ncl_entries` notebook in `ncl/ncl_index/ncl_entries.rst`
- A new row in `ncl/ncl_index/ncl-index-table.csv` for each NCL function
- A new receipts entry under `ncl/receipts/`
- A new raw `.ncl` script to store `NCL Code` for receipt within `ncl/ncl_raw/`

### Open a New Pull Request

> A Pull Request is a request to merge code from your fork of `geocat-applications`

Reviewer 2: ```suggestion
A pull request is a request to merge code from your fork of `geocat-applications`
```

on GitHub to the main repository. Project Pythia has extensive
[pull request guides and documentation](https://foundations.projectpythia.org/foundations/github/github-pull-request.html)
if you'd like more information

When opening a pull request, if you want to open a pull request but are not ready for it
to be reviewed, you can open the pull request as a draft. This is also a good way to get
feedback on your work that might not be ready to contribute yet.

> When a new Pull Request is created a deployment preview will be generated and added as a

Reviewer 2: ```suggestion
When a new pull request is created a deployment preview will be generated and added as a
```

comment.

![deploy_preview](https://github.com/NCAR/geocat-applications/assets/22159116/b42ace34-96d6-47b2-8c85-f0c20fa71927)

Follow the link to view and confirm your changes are being generated as expected. The preview
can take a few minutes to update, but any new changes made to the branch will be automatically

> added to the Pull Request. The deployment preview link will also be update to display the most

Reviewer 2: ```suggestion
added to the pull request. The deployment preview link will also be update to display the most
```

recent preview on a draft `geocat-applications` website.

Follow the deployment preview link to view and confirm your changes are being generated as expected.
The preview can take a few minutes to generate, and will update to include the most recent changes
on the PR.

### Address Feedback
After you open your pull request, the GeoCAT team will review it and may provide feedback like asking
for changes or suggesting improvements. You can address this feedback by making changes to your
branch and pushing them to your fork. The pull request and deployment preview comment will
automatically update with your changes.

The GeoCAT team appreciates your contributions and will do our best to review your pull request in
a timely manner. It is totally normal to have to make several rounds of changes to your pull request
before it is ready to be merged, especially if you are new to the project.

Once your pull request is approved by a core maintainer and passes the relevant checks, it will be merged into the
main repository!

### Delete Branch

> Once the Pull Request is closed and merged you can [delete your working branch](https://foundations.projectpythia.org/foundations/github/git-branches.html#deleting-branches).

Reviewer 2: ```suggestion
Once the pull request is closed and merged you can [delete your working branch](https://foundations.projectpythia.org/foundations/github/git-branches.html#deleting-branches).
```

the purpose of adding an entry to the NCL Index without the need for a full
**NCL Entry**. These files are accessible from the geocat-applications webpage,
but are not listed on TOCs or intended to be read as standalone or
comprehensive guides. They are intended to provide more extensive testing than is useful
in an NCL entry. Receipts are intended to be the minimal amount of documentation necessary to
add an entry to the NCL Index in cases where a full **NCL Entry** is not
necessary or where providing an initial entry to the NCL Index is more important
than waiting for a full **NCL Entry** to be completed.

1. Create a new file in `ncl/ncl_receipts/` based off of the
   `templates/receipt_template.ipynb` template.

1. Create raw `.ncl` script within `ncl/ncl_raw/` to store `NCL Code` section

1. Remove placeholder and descriptive text below `NCL Code`,
`Python Functionality`, `Comparison`

1. Make sure to clear and run all outputs before asking for a review

1. Add a new line to the `ncl/ncl_index/ncl-index-table.csv` file

[Example of a NCL Receipt](https://ncar.github.io/geocat-applications/ncl/receipts/days_in_month.html)

### Generate the Documentation Locally

> At the base of the reposistory `geocat-applications/`, run

Reviewer 2: ```suggestion
From the `geocat-applications/` directory, run
```

```
make clean html
```
This will build the documentation locally so you can view your changes.

New documentation can be viewed by opening `index.html` generated under
 `_build/html/index.html` on a local browser:

```
# on Mac
open _build/html/index.html

# Otherwise, open with a specific browser, like Firefox
firefox _build/html/index.html
```

> Running this commands on the terminal will open a new browser tab with the

Reviewer 2: ```suggestion
Running these commands from the terminal will open a new browser tab with the
```

changes you have made to the repository. The webpage runs on your local host,
but all the pages are linked together and act as they will on the
`geocat-applications` website. You can browse your changes, navigate with links,
and view images as you would on the official website.

## Contribute the Code

> Once you have completed your changes on notebooks and they are ready for review

Reviewer 2: ```suggestion
Once you have made your changes on notebooks and they are ready for review
```

by the GeoCAT team, you can open a new Pull Request. This section describes how
to open a pull request and what to expect after you open it.

### Check Files Changed

For a Python Entry, the files changed should include:
- New Python page under relevant `applications` folder
- Link to notebook in `.rst` file in the same directory as `applications/` (for
example: if new notebook is part of `data_analysis`, add link to
`applications/data_analysis/data_analysis.rst`)
- Link new notebook in `applications/applications.rst` file

For a NCL to Python entry, the files changes should include:
- New NCL to Python notebook under `ncl/ncl_entries`
- Link to new `ncl_entries` notebook in `ncl/ncl_index/ncl_entries.rst`
- A new row in `ncl/ncl_index/ncl-index-table.csv` for each NCL function
- A new receipts entry under `ncl/receipts/`
- A new raw `.ncl` script to store `NCL Code` for receipt within `ncl/ncl_raw/`

> ### Open a New Pull Request

Reviewer 2: It looks like this might be missing the push changes section and there's a lot of learning content in here as well. Is this also a section we could lean on Project Pythia Foundations for?  Maybe we could do something like [GeoCAT-comp](https://geocat-comp.readthedocs.io/en/stable/contrib.html#contribute-your-code) with some additional links to Project Pythia Foundations for extra support.

```
make clean html
```
This will build the documentation locally.

New documentation can be viewed by opening `index.html` generated under
 `_build/html/index.html` on a local browser

```
firefox _build/html/index.html
```

## Contribute the Code
Once you have completed your changes on notebooks and they are ready for review
by the GeoCAT team, you can open a new Pull Request. This section describes how
to open a pull request and what to expect after you open it.

### Check Files Changed

For a Python Entry, the files changed should include:
- New Python page under `applications`
- Link to directory in `.rst` file in the same directory as `applications/`
- Link new file in `applications/applications.rst` file

For a NCL to Python entry, the files changes should include:
- A new receipts entry under `receipts`
- A NCL to Python notebook under `ncl_entries`
- A new row in `ncl-index-table.csv`

### Open a new Pull Request
A Pull Request is a request to merge code from your local fork to the main repository.
GitHub has extensive [pull request guides and documentation](https://docs.github.com/en/pull-requests)
if you'd like more information

> On your local fork of the `geocat-applications` repo:

Reviewer 2: Should this be something more like "on your fork of `geocat-applications` on GitHub"?

Specifically, the terminology "local fork" is a little confusing to me.  Are you describing how to open a PR in VS Code?  I don't know that we want to presume everyone will be using the same IDE.

I think just copying over the section from GeoCAT-comp could work fine here for this full section probably.

1. Open "Contribute" tab
2. Select "Open Pull Request"
3. Select "New Pull Request" button

This will open page to compare your local changes to the main repository. Make sure that
the `base repository` is `NCAR/geocat-applications` and the `base` branch is set to `main`.
This will generate a pull request comparing your local changes to the  main branch of the
main `geocat-applications` repository. The `head repository` and `compare` branch should
be your forked repo and the branch that contains your changes respectively.

The pull request form will display all the changes you've made on the branch compared to
the main repo. We recommend adding a short descriptive title to the pull request (e.g.
"Add Python datetime and days_in_month"). In the body of the pull request please list

> out the NCL or Python functions being covered with a short description of the notebooks.

Reviewer 2: Or maybe we remove it here and just add it to the PR template?  I don't know that folks need a warning that this will be requested.  Plus then we just have to maintain it in one spot that is closer to where it's being used.
